The big ClassLoader change

stuartwdouglas opened the pull request:
Fixes #5597 #1724 #5354 #770 #3917 #6652

This changes the way Quarkus ClassLoading works,
to allow for isolated class loaders.

It also unifies how Quarkus is launched, so every
different mode we support uses the same mechanism
for both curation and launch.

Tests are now run in an isolated ClassLoader, which
means that a proxy is created that runs the tests
from within the isolated ClassLoader. This currently
has a quirk where @BeforeAll methods are run twice,
which will be fixed in the next JUnit release. This
can be worked around by using @QuarkusBeforeAll.

Apparently JUnit 5.6 which will fix this should be released on the 20th according to https://github.com/junit-team/junit5/milestone/46, which will allow this restriction to be removed.

This still needs performance testing, but otherwise is very close to being complete.

All of these should be addressed in the new few days.

stuartwdouglas committed:
The big ClassLoader change

This changes the way Quarkus ClassLoading works,
to allow for isolated class loaders.

It also unifies how Quarkus is launched, so every
different mode we support uses the same mechanism
for both curation and launch.

Tests are now run in an isolated ClassLoader, which
means that a proxy is created that runs the tests
from within the isolated ClassLoader. This currently
has a quirk where @BeforeAll methods are run twice,
which will be fixed in the next JUnit release. This
can be worked around by using @QuarkusBeforeAll.

aloubyansky commented on core/deployment/src/main/java/io/quarkus/runner/bootstrap/GenerateConfigTask.java:
```diff
@@ -0,0 +1,161 @@
+
+            String existing = "";
+            if (Files.exists(configFile)) {
+                try (InputStream in = new FileInputStream(configFile.toFile())) {
```
Just curious, any reason not to do it the nio way with Files.newInputStream(configFile)?

aloubyansky commented on core/deployment/src/main/java/io/quarkus/runner/bootstrap/GenerateConfigTask.java:
```diff
@@ -0,0 +1,161 @@
+                sb.append("\n");
+            }
+
+            try (FileOutputStream out = new FileOutputStream(configFile.toFile(), true)) {
```
Same here. Or it looks like alternatively you could do `Files.newBufferedWriter` and write directly there instead of `sb` using `newLine()`.

aloubyansky reviewed (commented):
I haven't yet reviewed it completely. I'll probably do it in a few steps as it's quite large and not trivial.

geoand commented on test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java:
```diff
@@ -3,220 +3,90 @@
 
-        appCl = createQuarkusBuildClassLoader(appClassLocation);
-        originalCl = setCCL(appCl);
+            QuarkusBootstrap.builder(appClassLocation);
```
Not that it should be an issue, but I assume this is not needed since `QuarkusBootstrap.builder(appClassLocation).setMode(QuarkusBootstrap.Mode.TEST);` is being used later

geoand commented on docs/src/main/asciidoc/class-loading-reference.adoc:
```diff
@@ -0,0 +1,23 @@
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
+////
+= Quarkus - Class Loading Reference
```
+100 for this doc!

It would be wonderful to have this document completed so people that haven't followed the ClassLoader system design discussions can have a good reference point and perhaps even be able to contribute somewhat to the review process

stuartwdouglas commented:
This commit includes a class loading guide that gives a bit of an explanation as to how it all works, when reviewing this is probably a good place to start.

gsmet commented on docs/src/main/asciidoc/class-loading-reference.adoc:
```diff
@@ -0,0 +1,163 @@
+
+This loads all the `-deployment` artifacts and their dependencies, as well as other user dependencies. It does not load the
+application root or any hot deployed code. This ClassLoader is persistent, even if the application restarts it will remain
+(which is why it cannot load application classes that may be hot deployed). It's parent is the base ClassLoader, and it is
```
```suggestion
(which is why it cannot load application classes that may be hot deployed). Its parent is the base ClassLoader, and it is
```

gsmet commented on docs/src/main/asciidoc/class-loading-reference.adoc:
```diff
@@ -0,0 +1,163 @@
+This loads all the `-deployment` artifacts and their dependencies, as well as other user dependencies. It does not load the
+application root or any hot deployed code. This ClassLoader is persistent, even if the application restarts it will remain
+(which is why it cannot load application classes that may be hot deployed). It's parent is the base ClassLoader, and it is
+transformer safe.
```
I think we should have an explanation of what that means.

geoand replied:
The explanation exists, but it's further down, so I agree it should be moved up here

geoand replied:
Can you also please add an explanation why this classloader needs to be transformer safe? Is this need driven by the fact that we need to transform dependencies in some cases?

gsmet commented on docs/src/main/asciidoc/class-loading-reference.adoc:
```diff
@@ -0,0 +1,163 @@
+
+Deployment ClassLoader::
+
+This can load all application classes, it's parent is the Augment ClassLoader so it can also load all deployment classes.
```
```suggestion
This can load all application classes, its parent is the Augment ClassLoader so it can also load all deployment classes.
```

gsmet commented on docs/src/main/asciidoc/class-loading-reference.adoc:
```diff
@@ -0,0 +1,163 @@
+
+This ClassLoader is non-persistent, it will be re-created when the application is started, and is isolated. This ClassLoader
+is the context ClassLoader that is used when running the build steps. It is transformer safe so loading application classes
+will not prevent them from being transformed.
```
Oh, you explain it there.

gsmet commented on docs/src/main/asciidoc/class-loading-reference.adoc:
```diff
@@ -0,0 +1,163 @@
+
+This loads all the runtime extension dependencies, as well as other user dependencies. It does not load the
+application root or any hot deployed code. This ClassLoader is persistent, even if the application restarts it will remain
+(which is why it cannot load application classes that may be hot deployed). It's parent is the base ClassLoader.
```
```suggestion
(which is why it cannot load application classes that may be hot deployed). Its parent is the base ClassLoader.
```

gsmet commented on docs/src/main/asciidoc/class-loading-reference.adoc:
```diff
@@ -0,0 +1,163 @@
+values in a static field in the deployment side, and expect to read it at runtime. This allows dev and test applications
+to behave more like a production application (production applications are isolated in that they run in a whole new JVM).
+
+This also means that the runtime version can be linked against a different set of dependencies, e.g. the hibernate
```
```suggestion
This also means that the runtime version can be linked against a different set of dependencies, e.g. the Hibernate ORM
```

gsmet commented on docs/src/main/asciidoc/class-loading-reference.adoc:
```diff
@@ -0,0 +1,163 @@
+
+Runtime Class Loader::
+
+This ClassLoader is used to load the application classes and other hot deployable resources. It's parent it the base runtime
```
```suggestion
This ClassLoader is used to load the application classes and other hot deployable resources. Its parent is the base runtime
```

gsmet commented on docs/src/main/asciidoc/class-loading-reference.adoc:
```diff
@@ -0,0 +1,163 @@
+
+== Isolated ClassLoaders
+
+The runtime ClassLoader is always isolated. This means that it will have it's own copies of almost every class from the
```
```suggestion
The runtime ClassLoader is always isolated. This means that it will have its own copies of almost every class from the
```

gsmet commented on docs/src/main/asciidoc/class-loading-reference.adoc:
```diff
@@ -0,0 +1,163 @@
+- Quarkus bootstrap itself
+
+If this is required it can be configured in the `quarkus-bootstrap-maven-plugin`. Note that if you
+mark a dependency as parent first then all of it's dependencies must also be parent first,
```
```suggestion
mark a dependency as parent first then all of its dependencies must also be parent first,
```

geoand replied:
I think this needs to have more attention drawn to it. Extension authors would probably need to know more about this

gsmet commented on docs/src/main/asciidoc/class-loading-reference.adoc:
```diff
@@ -0,0 +1,163 @@
+There are some dependencies that we can be sure we do not want. This generally happens when a dependency has had a name
+change (e.g. smallrye-config changing groups from `org.smallrye` to `org.smallrye.config`, the `javax` -> `jakarta` rename).
+This can cause problems, as if these artifacts end up in the dependency tree out of date classes can be loaded that are
+not compatible with Quarkus. To deal with this this extensions can specify artifacts that should never be loaded. This is
```
```suggestion
not compatible with Quarkus. To deal with this, extensions can specify artifacts that should never be loaded. This is
```

gsmet commented on docs/src/main/asciidoc/class-loading-reference.adoc:
```diff
@@ -0,0 +1,163 @@
+</plugin>
+----
+
+This should only be done if the extension depends on a never version of these artifacts. If the extension does not bring
```
```suggestion
This should only be done if the extension depends on a newer version of these artifacts. If the extension does not bring
```

gsmet reviewed (dismissed):
I will dig more into this but here is a small review of the documentation part.

geoand commented on docs/src/main/asciidoc/class-loading-reference.adoc:
```diff
@@ -0,0 +1,163 @@
+
+Base Runtime ClassLoader::
+
+This loads all the runtime extension dependencies, as well as other user dependencies. It does not load the
```
`as well as other user dependencies` is mentioned here for `Base Runtime Classloader`, however `Augment ClassLoader` also mentions `as well as other user dependencies`.

This seems kind of confusing, but I assume that the former is done at build time and the latter at runtime? Maybe a note should be made about this?

geoand commented on docs/src/main/asciidoc/class-loading-reference.adoc:
```diff
@@ -0,0 +1,163 @@
+
+This loads code that is not hot-reloadable, but it does support transformation (although once the class is loaded this
+transformation is no longer possible). This means that only transformers registered in the first application start
+will take effect, however as these transformers are expected to be idempotent this should not cause problems.
```
Could we have an example of what sort of transformations are meant here?

geoand commented:
I took at this a little and I really love the way things get a lot cleaner

FroMage commented on integration-tests/hibernate-orm-panache/src/test/java/io/quarkus/it/panache/PanacheFunctionalityTest.java:
```diff
@@ -45,6 +55,7 @@ public void testPanacheSerialisation() {
 
     @DisabledOnNativeImage
     @Test
+    @ActivateRequestContext
```
Is the request context no longer active by default for tests, then?

stuartwdouglas replied:
ah, good catch. I need to fix this, the test scope setup is being run from outside the scope of the deployment so it is not actually activating the scope.

stuartwdouglas commented:
The windows tests are failing because something somewhere is not closing a file handle for the libs/$app}.jar file in the gradle build. The gradle tests can't delete their temp folders and fail as a result.

aloubyansky commented on core/deployment/src/main/java/io/quarkus/runner/bootstrap/AugmentActionImpl.java:
```diff
@@ -0,0 +1,214 @@
+        BuildResult result = runAugment(true, Collections.emptySet(), ArtifactResultBuildItem.class);
+        JarBuildItem jarBuildItem = result.consumeOptional(JarBuildItem.class);
+        NativeImageBuildItem nativeImageBuildItem = result.consumeOptional(NativeImageBuildItem.class);
+        curatedApplication.close();
```
I guess it'd be safer to close it in `finally` in case `runAugment` throws a runtime exception. In general though, I don't see `CuratedApplication` being closed anywhere else which looks like a major issue. Did I overlook something?
I get that depending on the mode it may have different lifecycles. But it'd be great to improve the design around it to make it more clear.
We do need to test prove it on Windows which is more sensitive to closing resources.

aloubyansky replied:
Ah, sorry, I see RunningQuarkusApplicationImpl.

stuartwdouglas replied:
There were some gradle failures on windows that were due to this (the app not being closed). The gradle tests create a temp dir and the test fails if the dir cannot be released.

aloubyansky commented on core/deployment/src/main/java/io/quarkus/runner/bootstrap/StartupActionImpl.java:
```diff
@@ -0,0 +1,172 @@
+
+    public StartupActionImpl(CuratedApplication curatedApplication, AugmentAction augmentAction, BuildResult buildResult) {
+        this.curatedApplication = curatedApplication;
+        this.augmentAction = augmentAction;
```
augmentAction looks like a leftover. Can be removed?

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/CuratedApplication.java:
```diff
@@ -0,0 +1,279 @@
+        return augmentClassLoader;
+    }
+
+    private AppArtifactKey getKey(AppDependency i) {
```
This method looks equivalent to `i.getArtifact().getKey()`.

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/AppModelGradleResolver.java:
```diff
@@ -228,4 +267,31 @@ private Properties resolveDescriptor(final Path path) {
         return rtProps;
     }
+
+    static final class AppKey {
```
Why not use AppArtifactKey instead?

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusTestConfig.java:
```diff
@@ -34,7 +33,7 @@ public void setupTest() {
             for (Test test : getProject().getTasks().withType(Test.class)) {
                 final Map<String, Object> props = test.getSystemProperties();
-                props.put(BootstrapClassLoaderFactory.PROP_DEPLOYMENT_CP, deploymentCp);
+                //props.put(BootstrapClassLoaderFactory.PROP_DEPLOYMENT_CP, deploymentCp);
```
That's an interesting one. If it's not needed any more, this whole task is questionable. We shouldn't be resolving the app model here, at a minimum.

aloubyansky replied:
IIUIC, here is a potential issue with this change. Gradle and Maven perform version convergence differently. The way it currently is in master we are using the same Gradle resolver for tests and actual builds.
With this change it looks like we are going to be using the Maven resolver for the tests and the Gradle one for the actual builds. And the dev mode will always be using the Maven resolver in `DevModeMain`.

stuartwdouglas replied:
This is something I am not sure about, and wanted some feedback on.

If this is removed we fall back to class path based discovery in the maven resolver (unless there happens to be a pom present as well). The main advantage of this is that it should allow gradle tests to also run in the IDE, which does not work at the moment. 

If that is not possible then we need to somehow encode the app model into this system prop and restore the old behaviour.

aloubyansky replied:
While it does not look elegant in this case, we are ending up testing against a different set of deps than the ones used for the build. They should be aligned. Either we use those resolved by Gradle all the way or we use those that we resolve in QuarkusBootstrap.

aloubyansky replied:
By not looking elegant I refer to the classpath shared through the system property.

aloubyansky commented on devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java:
```diff
@@ -528,49 +528,45 @@ void prepare() throws Exception {
+            MavenArtifactResolver mavenArtifactResolver = MavenArtifactResolver.builder()
+                    .setRepositorySystem(repoSystem)
+                    .setRepositorySystemSession(repoSession)
+                    .setRemoteRepositories(repos)
```
I think `.setWorkspace(localProject.getWorkspace())` should still be necessary. Otherwise, it won't be able to resolve artifacts that are supposed to be produced by the workspace but haven't actually been installed.

aloubyansky replied:
Actually, you don't need to set the workspace here since you are not resolving any app deps with this resolver. Which means you can also simply use `DevMojo.this.repoSystem` and remove `repoSystem` local variable.

aloubyansky commented on devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java:
```diff
@@ -528,49 +528,45 @@ void prepare() throws Exception {
+                    .setRepositorySystemSession(repoSession)
+                    .setRemoteRepositories(repos)
+                    .build();
+            DependencyResult cpRes = mavenArtifactResolver.resolveManagedDependencies(
```
I'd simply use `mavenArtifactResolver.resolveDependencies(` instead. It does delegate to this method after all, it's purely to use softer words :) and possibly create less confusion for those reading the code and not to learn what these method names actually mean.

aloubyansky commented on core/devmode/src/main/java/io/quarkus/dev/DevModeMain.java:
```diff
@@ -43,24 +29,16 @@
-        Timing.staticInitStarted();
 
         try (InputStream devModeCp = DevModeMain.class.getClassLoader().getResourceAsStream(DEV_MODE_CONTEXT)) {
             DevModeContext context = (DevModeContext) new ObjectInputStream(new DataInputStream(devModeCp)).readObject();
```
`context` doesn't seem to be closed, btw.

stuartwdouglas replied:
context is not Closeable

aloubyansky replied:
Sorry, I meant the instance of DevModeMain.

aloubyansky replied:
instantiated below at line 45

aloubyansky replied:
So what about closing the instance of DevModeMain?

stuartwdouglas replied:
I have changed it to be closed after the park, but in practice it would have been cleaned up by the normal shutdown hook anyway.

aloubyansky commented on devtools/maven/src/main/java/io/quarkus/maven/GenerateConfigMojo.java:
```diff
@@ -95,48 +91,28 @@ public void execute() throws MojoExecutionException {
             throw new MojoExecutionException("No resources directory, cannot create application.properties");
         }
+        try {
+            CuratedApplication curatedApplication = QuarkusBootstrap.builder(Paths.get(project.getBuild().getOutputDirectory()))
```
Looks like `curatedApplication` isn't closed. I would also strongly recommend creating maven artifact resolvers in mojos and passing them to the bootstrap.

stuartwdouglas replied:
how do you do that? This is a seperate JVM. 

One option could be to resolve the app model in the mojo and serialize the resolved app.

aloubyansky replied:
Isn't curated application going to be usually created in the Mojo when running from maven?

stuartwdouglas replied:
Except for dev mode, and tests (which have always done their own resolution). I can make dev mode work by serialising the result into the dev mode runner jar, but there is not much I can do with tests.

aloubyansky commented on devtools/maven/src/main/java/io/quarkus/maven/NativeImageMojo.java:
```diff
@@ -186,278 +175,157 @@ public NativeImageMojo() {
+                System.setProperty(e.getKey(), e.getValue());
+            }
+            try {
+                CuratedApplication curatedApplication = QuarkusBootstrap.builder(outputDirectory.toPath())
```
Let's create the maven artifact resolver and pass it to the bootstrap here.

stuartwdouglas replied:
Ok. All the resolution stuff is what I am least sure about, I think it is likely that some of my changes are not 100% correct, or at the very least not as efficient as they could be.

aloubyansky replied:
First, the repo system, session, repos, etc here are properly initialized in the mojo, let's call them the original ones, because in tests and now in bootstrap we are re-creating them (currently, most probably using different versions of the api and impl and there have been lots of issues related to this initialization). Second, avoiding them being re-initialized is going to be a lot more efficient, given that Maven has already cached some artifact and resolution request-related data.

stuartwdouglas commented:
CI is now passing

aloubyansky commented on devtools/maven/src/main/java/io/quarkus/maven/NativeImageMojo.java:
```diff
@@ -186,278 +176,164 @@ public NativeImageMojo() {
     @Override
     public void execute() throws MojoExecutionException, MojoFailureException {
 
-        if (project.getPackaging().equals("pom") && appArtifact == null) {
```
Was there a reason to remove `appArtifact` check here?
`appArtifact` is used in the platform integration tests. The reason I allowed `pom` project and `appArtifact` combination was that the project was simply pulling all the code from the plugin config and dependencies and didn't produce any JAR.

aloubyansky replied:
It looks like somebody has removed pom packaging from the platform tests, although the rest of the config remains the same. It could still be just a POM though.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/BootstrapAppModelFactory.java:
```diff
@@ -0,0 +1,512 @@
+                try (JarInputStream file = new JarInputStream(jarUrl.openConnection().getInputStream())) {
+                    JarEntry entry = file.getNextJarEntry();
+                    while (entry != null) {
+                        if (entry.getName().endsWith("/pom.properties") && entry.getName().startsWith("META-INF/maven")) {
```
The issue is that this is not very reliable. Neither `pom.properties` nor `pom.xml` are always present. I don't think JARs built with Gradle contain those. You can try a simple test: scan JARs in your local repo and list those that don't include `META-INF/maven`.
So this approach can be used only as "the best effort" in cases when nothing else is available.

stuartwdouglas replied:
Thinking about it we don't actually need the artifact names here, the path is what is really important. We just need some way to make sure we don't get jdk jars.

aloubyansky replied:
Right, unless we wanted to apply a filter.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/BootstrapAppModelFactory.java:
```diff
@@ -0,0 +1,512 @@
+        return project.getOutputDir().resolve(QUARKUS).resolve(BOOTSTRAP).resolve(DEPLOYMENT_CP);
+    }
+
+    private static org.apache.maven.model.RepositoryPolicy toMavenRepoPolicy(RepositoryPolicy policy) {
```
Looks like this one isn't used anywhere.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:
```diff
@@ -0,0 +1,504 @@
+            }
+        }
+        if (!banned) {
+            if(resources.isEmpty() || aggregateParentResources) {
```
This one looks somewhat confusing. If `aggregateParentResources` is false then it will still be looking for the resources in the parent but only if nothing was found locally. Is there a specific use-case where this is needed? I'm just thinking wouldn't simply consulting the parent only if `aggregateParentResources` was true work?

stuartwdouglas replied:
this is used for the deployment and runtime class loaders, so that getResources gets everything that would be on the actual application class path, without anything that may have leaked in on the system class path.

I am not sure about just never consulting the parent, as it would effectively hide JDK resources. Arc needs to be able to index JDK classes at times, so if we did not delegate at all this would cause problems.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:
```diff
@@ -0,0 +1,504 @@
+                return res.getUrl();
+            }
+        }
+        return parent.getResource(nm);
```
How about taking the value of `aggregateParentResources` into account here and in `getResourceAsStream` below? Wondering whether it's this way on purpose. Otherwise, it'd be probably a good idea to have a consistent resource access switch.

stuartwdouglas replied:
So we can't completely hide the JDK resources or ArC can't index them, but we want to hide duplicate/unwanted copies of META-INF/services files. I was not 100% sure about this but I can't really think of a better approach.

aloubyansky replied:
Ok

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/BootstrapAppModelResolver.java:
```diff
@@ -130,24 +143,42 @@ public AppModel resolveModel(AppArtifact appArtifact, List<AppDependency> direct
+        } else {
+            //if there is no main artifact we assume we already have all the deps we need
+            //we just turn them into a DependencyNode
+            resolvedDeps = mvn.toDependencyTree(directMvnDeps, managedRepos).getRoot();
```
`directMvnDeps` are supposed to be only the direct deps, their transitive deps aren't in there.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/BootstrapAppModelResolver.java:
```diff
@@ -157,28 +188,38 @@ public boolean visitEnter(DependencyNode node) {
         List<AppDependency> deploymentDeps = Collections.emptyList();
-        if(deploymentInjector.isInjectedDeps()) {
-            final DependencyGraphTransformationContext context = new SimpleDependencyGraphTransformationContext(mvn.getSession());
+        List<Dependency> deploymentDependencies = new ArrayList<>();
```
Looks like `deploymentDependencies` is not used for anything useful.

aloubyansky replied:
Please remove `deploymentDependencies`.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/BootstrapAppModelResolver.java:
```diff
@@ -336,4 +418,31 @@ private static AppArtifact toAppArtifact(Artifact artifact) {
         return directMvnDeps;
     }
+
+    static final class AppKey {
```
Let's keep using `AppArtifactKey` instead.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/BootstrapAppModelResolver.java:
```diff
@@ -297,37 +352,64 @@ private VersionRangeResult resolveVersionRangeResult(AppArtifact appArtifact, St
         return appArtifact;
     }
 
+    private static AppArtifact toAppArtifact(Artifact artifact, Map<AppKey, AppArtifact> versions) {
```
Ah, this method is actually not used. Is it? So AppKey can simply be removed with this method.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/DeploymentInjectingDependencyVisitor.java:
```diff
@@ -111,19 +115,23 @@ private void processPlatformArtifact(DependencyNode node, Path descriptor) throw
+        node.setData(QUARKUS_DEPLOYMENT_ARTIFACT, deploymentArtifact);
+        runtimeNodes.add(node);
+        managedDeps.add(new Dependency(node.getArtifact(), JavaScopes.COMPILE));
+        runtimeExtensionDeps.add(new Dependency(node.getArtifact(), JavaScopes.COMPILE));
```
I guess it's safe to re-use `new Dependency(node.getArtifact(), JavaScopes.COMPILE)` created above.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenArtifactResolver.java:
```diff
@@ -312,6 +317,46 @@ public DependencyResult resolveManagedDependencies(Artifact artifact, List<Depen
+        }
+        return result;
+    }
+    public DependencyResult resolveManagedDependencies(List<Dependency> deps, List<Dependency> managedDeps, List<RemoteRepository> mainRepos) throws AppModelResolverException {
```
This method doesn't seem to be used anywhere. Let's remove it for now.

aloubyansky replied:
Please remove this method.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenArtifactResolver.java:
```diff
@@ -328,7 +373,7 @@ private CollectRequest newCollectManagedRequest(Artifact artifact, List<Dependen
         Collection<String> excluded;
         if(excludedScopes.length == 0) {
-            excluded = Arrays.asList(new String[] {"test", "provided"});
+            excluded = Collections.emptyList();
```
I agree this was an ugly choice to design it this way. But by default "test" and "provided" should be excluded. That is, if the caller doesn't provide anything.

stuartwdouglas replied:
This is why there is now explicit devmode and test flags, as if you are running the dev mode tests you need both.

stuartwdouglas replied:
That is why I changed this, as there are cases where there are no excluded scopes, so now it is up to the caller.

aloubyansky replied:
This is a different default behavior though. But let's leave it for now. I didn't do it well in the first place, so I'll probably review it later.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/options/BootstrapMavenOptions.java:
```diff
@@ -90,7 +93,7 @@ public String getOptionValue(String name) {
     public String[] getOptionValues(String name) {
         final Object o = options.get(name);
-        return o == null ? null : (String[]) o;
+        return o == null ? null : o instanceof String ? new String[] { o.toString() } : (String[]) o;
```
Just curious, was there a case when it actually was a `String`? Which option was that?

stuartwdouglas replied:
I think it was these ones: https://github.com/quarkusio/quarkus/blob/8f0d1587609fd1d055caaa92bce99f70998b20b3/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/options/BootstrapMavenOptionsParser.java#L25

aloubyansky replied:
If you recall how to reproduce that, please let me know. If not we'll merge it anyway. But I'd like to see the stacktrace.

stuartwdouglas replied:
I can't actually remember, and it might not even be a problem any more as this PR has gone through so many changes.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/util/IoUtils.java:
```diff
@@ -1,6 +1,7 @@
 package io.quarkus.bootstrap.util;
 
 import java.io.BufferedReader;
+import java.io.ByteArrayOutputStream;
```
This one is redundant.

aloubyansky commented on independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/CollectDependenciesBase.java:
```diff
@@ -43,8 +44,8 @@ public void testCollectedDependencies() throws Exception {
-        final List<AppDependency> resolvedDeps = getTestResolver().resolveModel(root.toAppArtifact()).getAllDependencies();
-        assertEquals(expected, resolvedDeps);
+        final List<AppDependency> resolvedDeps = getTestResolver().resolveModel(root.toAppArtifact()).getFullDeploymentDeps();
+        assertEquals(new HashSet<>(expected), new HashSet<>(resolvedDeps));
```
Why? I think it's important to keep the list.

aloubyansky commented on independent-projects/bootstrap/core/pom.xml:
```diff
@@ -14,6 +14,10 @@
+        <dependency>
+            <groupId>org.ow2.asm</groupId>
+            <artifactId>asm</artifactId>
+        </dependency>
```
Where is this used in core?

stuartwdouglas replied:
The ClassLoader implementation allows for transformers, and the transformer API is based on ASM.

aloubyansky commented on integration-tests/amazon-lambda/src/test/java/io/quarkus/it/amazon/lambda/AmazonLambdaSimpleTestCase.java:
```diff
@@ -29,9 +28,9 @@ public void testSimpleLambdaFailure() throws Exception {
+        } catch (Exception e) {
             Assertions.assertEquals(ProcessingService.CAN_ONLY_GREET_NICKNAMES, e.getMessage());
-            Assertions.assertEquals(IllegalArgumentException.class.getName(), e.getType());
+            //Assertions.assertEquals(IllegalArgumentException.class.getName(), e.getType());
```
TBH, I'm not familiar with this test but if for LambdaException it's still relevant to make sure the type is IAE, shouldn't this turn into multiple catches?

aloubyansky commented on test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java:
```diff
@@ -84,10 +84,10 @@ public void stop() {
+            ConfigProviderResolver cpr = ConfigProviderResolver.instance();
             cpr.releaseConfig(cpr.getConfig());
-        } catch (IllegalStateException ignored) {
+        } catch (Throwable ignored) {
```
`Exception` won't be generic enough here?

stuartwdouglas replied:
This may not be nessesary any more, there were some smallrye class loading problems that resulted in Error's being thrown. I think they are fixed now as the CL always returns null from getParent, but I need to re-check.

aloubyansky commented on test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java:
```diff
@@ -3,259 +3,123 @@
-        appCl = createQuarkusBuildClassLoader(appClassLocation);
-        originalCl = setCCL(appCl);
+            System.err
+                    .println("INIT TIME " + (System.currentTimeMillis() - ManagementFactory.getRuntimeMXBean().getStartTime()));
```
Is this a leftover?

aloubyansky commented on test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java:
```diff
@@ -3,259 +3,123 @@
+            CuratedApplication curatedApplication = runnerBuilder.setTest(true).build().bootstrap();
+            Timing.staticInitStarted(curatedApplication.getBaseRuntimeClassLoader());
+            System.err.println(
+                    "CURATE TIME " + (System.currentTimeMillis() - ManagementFactory.getRuntimeMXBean().getStartTime()));
```
This one too?

aloubyansky commented on test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java:
```diff
@@ -3,259 +3,123 @@
-                                Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> functions) {
-                            ClassLoader main = Thread.currentThread().getContextClassLoader();
+            System.err.println(
+                    "REAL START TIME " + (System.currentTimeMillis() - ManagementFactory.getRuntimeMXBean().getStartTime()));
```
And this one?

aloubyansky reviewed (commented):
@stuartwdouglas have you tried the platform TS with this PR? https://github.com/quarkusio/quarkus-platform

aloubyansky commented on core/deployment/src/main/java/io/quarkus/deployment/QuarkusAugmentor.java:
```diff
@@ -141,6 +145,9 @@ public BuildResult run() throws Exception {
             }
+            if (deploymentClassLoader instanceof Closeable) {
+                ((Closeable) deploymentClassLoader).close();
+            }
```
Just wondering whether `classLoader` which is the application's augment classloader should also be closed here.

stuartwdouglas replied:
No, they have different lifecycles. This one is recreated on each hot reload, the other lasts for the life of dev mode apps.

aloubyansky commented on core/deployment/src/main/java/io/quarkus/runner/bootstrap/GenerateConfigTask.java:
```diff
@@ -0,0 +1,157 @@
+                @Override
+                public void accept(BuildExecutionBuilder buildExecutionBuilder) {
+                    try {
+                        buildExecutionBuilder.produce(new ArchiveRootBuildItem(Files.createTempDirectory("empty")));
```
Shouldn't we `deleteOnExit` this dir? Or do a more controlled cleanup? Perhaps we should consider introducing some kind of Quarkus-post-build-clean-up tasks in perspective.

stuartwdouglas replied:
good catch, but in this case it can just be deleted immediately after.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/BootstrapAppModelFactory.java:
```diff
@@ -0,0 +1,522 @@
+                if (url.getProtocol().equals("jar")) {
+                    String path = url.getPath();
+                    if (path.startsWith("file:")) {
+                        path = path.substring(5, path.lastIndexOf('!'));
```
The code here is assuming it's a JAR. It could be a directory containing classes, couldn't it?

stuartwdouglas replied:
no, see the if statement above.

aloubyansky replied:
Ok, my point was though about directories containing classes that are on classpath. If you are launching from an IDE that is what should appear on the classpath, right? So why is this method adding only the JARs?

stuartwdouglas replied:
I need to think about this one a bit more, but it is really a last resort case when we have no info. I could just delete it but I think it is worth keeping and trying to solve any reported problems as they arise.

aloubyansky replied:
I think you can keep it for now. AFAIU, this method may get called when (re)building an app from a JAR in the first place. Probably a pretty good chance that everything comes in a JAR at that point.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/BootstrapAppModelFactory.java:
```diff
@@ -0,0 +1,522 @@
+                }
+            }
+
+            //we now have our deployment time artifacts, lets resolve all their deps
```
Aren't these the runtime artifacts?

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/BootstrapAppModelFactory.java:
```diff
@@ -0,0 +1,522 @@
+            }
+
+            //we now have our deployment time artifacts, lets resolve all their deps
+            AppModel model = resolver.resolveManagedModel(appArtifact, artifacts, managingProject);
```
For this invocation we'd need the actual GAVs.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/BootstrapAppModelFactory.java:
```diff
@@ -0,0 +1,522 @@
+    private CurationResult createAppModelForJar(Path appArtifactPath) {
+        log.debug("provideOutcome depsOrigin=" + dependenciesOrigin + ", versionUpdate=" + versionUpdate
+                + ", versionUpdateNumber="
+                + versionUpdateNumber);
```
Could you please change it to `.log.debugf(...)` instead?

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/BootstrapAppModelFactory.java:
```diff
@@ -212,6 +213,18 @@ public AppModelResolver getAppModelResolver() {
+        if (test) {
+            //gradle tests encode the result on the class path
+
+            try (InputStream existing = getClass().getResourceAsStream(BootstrapConstants.SERIALIZED_APP_MODEL)){
```
Wouldn't it be safer and perhaps more efficient to look for that file in `appClasses`? Now that the file is created in the classes dir, it will end up in the jar, right? Just wondering where there could be multiple files on the classpath.

stuartwdouglas replied:
It is created in the test class directory, so it should not end up in the jar. It does stay on the file system though so if you attempt to run a test from the IDE after a build it should still be picked up.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:
```diff
@@ -274,8 +274,17 @@ public InputStream getResourceAsStream(String nm) {
+                        //if we are already the initiating loader we might want to load this from the parent
+                        if (!parentFirst) {
+                            return parent.loadClass(name);
+                        }
```
Is there a test or specific use-case this is fixing in our build or TS?

stuartwdouglas replied:
This fixes some issues in the camel test suite that I am still investigating.

stuartwdouglas replied:
I have removed this. The issue in Camel is that they are packaging parts of XML parsers, e.g. a jar contains a copy of QName but not the parser that is using it. This causes problems as the parser uses the JDK version of QName but the app sees the copy in the isolated ClassLoader. 

Once this is in I am planning on working with the camel team to resolve it.

gastaldi commented on bom/runtime/pom.xml:
```diff
@@ -114,7 +114,7 @@
         <shrinkwrap.version>1.2.6</shrinkwrap.version>
         <rest-assured.version>4.1.2</rest-assured.version>
-        <junit.jupiter.version>5.5.2</junit.jupiter.version>
+        <junit.jupiter.version>5.6.0-RC1</junit.jupiter.version>
```
https://github.com/quarkusio/quarkus/pull/6663 updates to 5.6.0. It would be nice to rebase once that PR is merged

stuartwdouglas commented:
@aloubyansky I believe I have fixed all the issues you brought up. 

@gsmet I have fixed the docs issues. I have tested against the platform test suite, and there are some issues that will likely need to be fixed on the camel side (e.g. the XML parsing issue I mention above). Once this is in I will work with them to get this resolved.

aloubyansky commented on core/deployment/src/main/java/io/quarkus/runner/bootstrap/GenerateConfigTask.java:
```diff
@@ -0,0 +1,154 @@
+            Files.createDirectories(configFile.getParent());
+            Files.write(configFile, sb.toString().getBytes(StandardCharsets.UTF_8),
+                    Files.exists(configFile) ? new OpenOption[] { StandardOpenOption.APPEND } : new OpenOption[] {});
+            Files.deleteIfExists(temp);
```
Shouldn't it be deleted in the `finally`?

aloubyansky commented on core/deployment/src/main/java/io/quarkus/runner/bootstrap/StartupActionImpl.java:
```diff
@@ -0,0 +1,170 @@
+                            classWriter.write(i.getClassData());
+                        }
+                        log.infof("Wrote %s", classFile.getAbsolutePath());
+                    } catch (Throwable t) {
```
Does it really have to be a Throwable? Swallowing throwables in a loop may go pretty bad.

aloubyansky reviewed (approved):
Let's get this in, ASAP!

geoand commented:
Damn CI is acting up again...

stuartwdouglas commented:
I made another modification today as I found a class loader leak, if this passed it should be good to merge

stuartwdouglas commented:
CI seems to be working again, @gsmet if it passes can you hit the button?

gsmet commented:
Sure.

gsmet commented:
I rebased to trigger a new CI run... This CI issue is a nightmare :/
